### ISSUE TITLE
product-map is 10x slower than cartesian-map reduce

### ISSUE THREAD

**Issue by erg** [factor/factor#2962]
and `product-map` is hard to find compared to `cartesian-product` but it is the generalized version

```factor
: sequence-cartesian-product ( seqs -- seqs' )
    dup length 1 <= [
        [ [ 1array ] map ] map concat
    ] [
        2 cut [ first2 cartesian-product concat ] dip swap
        [ [ suffix ] cartesian-map concat ] reduce
    ] if ;

10,000 [ { { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 }  } 
sequence-cartesian-product drop ] times
1.8s

10,000 [ { { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 }  } 
<reversed> [ reverse! ] product-map drop ] times
10.8s
```

**Comment by mrjbq7** [comment]
You can’t test like that, it ends up using the non optimized compiler. You have to put that in a word and call it On Apr 8, 2024, at 11:03 PM, Doug Coleman ***@***.***> wrote:﻿
and product-map is hard to find compared to cartesian-product but it is the generalized version
: sequence-cartesian-product ( seqs -- seqs' )
    dup length 1 <= [
        [ [ 1array ] map ] map concat
    ] [
        2 cut [ first2 cartesian-product concat ] dip swap
        [ [ suffix ] cartesian-map concat ] reduce
    ] if ;

10,000 [ { { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 }  } 
sequence-cartesian-product drop ] times
1.8s

10,000 [ { { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 }  } 
<reversed> [ reverse! ] product-map drop ] times
10.8s

**Comment by mrjbq7** [comment]
I want to look later when I’m back home but I wonder if all the reversing is the overhead On Apr 9, 2024, at 7:11 AM, Doug Coleman ***@***.***> wrote:﻿
i think both compilers should be fast... here's it compiled with 10,000 as in
: test1 (  n -- ) [ { { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 }  } 
sequence-cartesian-product drop ] times ;

Running time: 1.793880458 seconds


IN: scratchpad auto-use : test2 ( n -- ) [ { { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 } { 1 2 3 4 5 }  } 
<reversed> [ reverse! ] product-map drop ] times ;

Running time: 6.0104585 seconds

**Comment by mrjbq7** [comment]
Fixed in 136836f8b0bc7adb75e5512f4e58cb773478582d
